File: python/grade-school/grade_school.py
import functools
""""Grade School Exercise"""

class School:
    """Grade School Student Roster"""

    def __init__(self):
        """setup the school data structure.
           will be a list of students where each student is a tuple.
           the first item of the tuple is the student name, the second is their grade.
        """
        self.students = []
        self.add_status = []

    def add_student(self, name, grade):
        """ add a student into a given grade into the school roster. """
        if name not in self.roster():
            self.students.append((name, grade))
            self.add_status.append(True)
        else:
            self.add_status.append(False)
    
    def added(self):
        """ return the status of additions """
        return self.add_status

    def roster(self):
        """ return a list of students in the school. Empty list if no students
            list should be sorted by grade then by name. 
        """
        return [item[0] for item in sorted(self.students, key=functools.cmp_to_key(self.compare))]

    def grade(self, grade_number):
        """ return an alphabetically sorted list of students with a given grade number.
            empty list if the grade has no students. 
        """
        return [item[0] for item in sorted(self.students) if item[1] == grade_number]

    def compare(self, item1, item2):
        """ compare two tuples and say which is earlier in the sort.
            returns -1 for item1 < item2, 0 if item1 == item2, and 1 if item1 > item2
        """

        # first compare by grade.
        if item1[1] == item2[1]:
            # same grade, so
            # compare by name
            if item1[0] == item2[0]:
                # same grade and name 1 == name 2
                return 0
            elif item1[0] < item2[0]:
                # name 1 < name 2
                return -1            
            # only greater than remains
            return 1
        elif item1[1] < item2[1]:
            # less than
            return -1
        # only greater than remains
        return 1

File: go/grade-school/grade_school.go
package gradeschool

import (
	"maps"
	"slices"
)

type School map[int][]string

// Create and return a reference to a new school
func New() *School {
	return &School{}
}

// Check if we have a student of a given name in any grade.
// @param student: Name of the student we are searching for
// @returns: True if we have a student with that name.
func (s *School) HasStudent(student string) bool {
	for grade := range maps.Keys((*s)) {
		if slices.Contains((*s)[grade], student) {
			return true
		}
	}
	return false
}

// Add a student to the roster
// @param student: Name of the student
// @param grade: The grade to place them into
// @returns: True if added to the roster
func (s *School) Add(student string, grade int) bool {
	if s.HasStudent(student) {
		return false
	}
	_, ok := (*s)[grade]
	if !ok {
		(*s)[grade] = make([]string, 0)
	}
	(*s)[grade] = append((*s)[grade], student)
	return true
}

// Return a sorted slice of student names for the given grade
// @param level: The grade level to return (empty result if doesn't exist)
// @returns: Slice of students in that grade in sorted order
func (s *School) Grade(level int) []string {
	ret := make([]string, len((*s)[level]))
	val, ok := (*s)[level]
	if !ok {
		return ret // Empty slice if level doesn't exist
	}
	copy(ret, val)
	slices.Sort(ret)
	return ret
}

// Return all students sorted first by grade then by name (ascending)
// @returns: Slice of sorted student names
func (s *School) Enrollment() []string {
	keys := make([]int, len(*s))
	copy(keys, slices.Collect(maps.Keys(*s)))
	slices.Sort(keys)
	ret := []string{}
	for _, key := range keys {
		ret = slices.Concat(ret, s.Grade(key))
	}
	return ret
}
